pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -471,7 +477,16 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 			return &status, nil
 		}
 		// Enforce the StatefulSet invariants
-		if identityMatches(set, replicas[i]) && storageMatches(set, replicas[i]) {
+		retentionMatch := true
+		if utilfeature.DefaultFeatureGate.Enabled(features.StatefulSetAutoDeletePVC) {
+			var err error
+			retentionMatch, err = ssc.podControl.ClaimsMatchRetentionPolicy(updateSet, replicas[i])
+			// An error is expected if the pod is not yet fully updated, and so return is treated as matching.
+			if err != nil {
```

Comment:
could you give more details about this logic here?

Reply:
We're checking if the retention policies are correct, if not there has been an update and we continue. This is parallel to the identityMatches and storageMatches checks.